src/Vulkan/RHI/VkAllocator.cpp

```diff
@@ -1,16 +1,127 @@
 #include "VkAllocator.h"
 
+#include <utility>
+
+#include <Vex/Resource.h>
+
+#include <Vulkan/VkDebug.h>
+#include <Vulkan/VkErrorHandler.h>
+#include <Vulkan/VkGPUContext.h>
+
 namespace vex::vk
 {
 
-void VkAllocator::OnPageAllocated(PageHandle handle, HeapType heapType)
+namespace AllocatorUtils
 {
-    VEX_NOT_YET_IMPLEMENTED();
+
+::vk::MemoryPropertyFlags GetMemoryPropsFromLocality(ResourceMemoryLocality locality)
+{
+    using enum ::vk::MemoryPropertyFlagBits;
+
+    switch (locality)
+    {
+    case ResourceMemoryLocality::GPUOnly:
+        return eDeviceLocal;
+    case ResourceMemoryLocality::CPURead:
+        return eHostCoherent | eHostVisible | eHostCached;
+    case ResourceMemoryLocality::CPUWrite:
+        return eHostCoherent | eHostVisible;
+    default:
+        VEX_LOG(Fatal, "Unable to deduce memory properties from locality");
+    }
+
+    std::unreachable();
+}
+
+u32 GetBestSuitedMemoryTypeIndex(::vk::PhysicalDevice device, u32 typeFilter, ::vk::MemoryPropertyFlags flags)
+{
+    ::vk::PhysicalDeviceMemoryProperties memProperties = device.getMemoryProperties();
+
+    for (u32 i = 0; i < memProperties.memoryTypeCount; i++)
+    {
+        if ((typeFilter & (1 << i)) && (memProperties.memoryTypes[i].propertyFlags & flags) == flags)
+        {
+            return i;
+        }
+    }
+
+    VEX_LOG(Fatal, "Unsuitable memory found for flags {:x}", static_cast<u32>(flags))
+    return 0;
+}
+
+} // namespace AllocatorUtils
+
+VkAllocator::VkAllocator(NonNullPtr<VkGPUContext> ctx)
+    : RHIAllocatorBase(ctx->physDevice.getMemoryProperties().memoryTypeCount)
```

Comment:
I'm guessing the reason that the memory type count is now dynamic (and no longer correlated to the HeapType enum) is due to vulkan rigidity versus potentially weird hardware?